### ISSUE TITLE
TP adjustments

### ISSUE THREAD

**Issue by insoochung** [insoochung/rtl-semantics-design-verification#42]
- [x] Calculate gradients for TP
- [x] Add design2vec flow, which uses gradient to adjust TP.
- [x] Generate yaml from adjusted TP

**Comment by insoochung** [comment]
- Create `nn/suggest_tp.py` which takes as inputs:
  - test yaml file
  - a coverpoint that we want to hit
- The script should modify the test parameters and create a yaml file.

**Comment by insoochung** [comment]
Commit ID: 147778d67990a34168f573568225fcfa93ad5f9e

```python
--- a/nn/layers.py
+++ b/nn/layers.py
@@ -83,8 +83,9 @@ class CdfgReader(tf.keras.layers.Layer):
   def call(self, inputs):
     # cdfg_xs (batch_size, num_nodes, num_features)
     # cdfg_as: (batch_size, num_nodes, num_nodes)
-    cdfg_xs = tf.gather_nd(self.cdfg_xs, inputs["graph"])
-    cdfg_as = tf.gather_nd(self.cdfg_as, inputs["graph"])
+    # assert 0, (self.cdfg_xs.shape, inputs["graph"].shape)
+    cdfg_xs = tf.gather(self.cdfg_xs, inputs["graph"])
+    cdfg_as = tf.gather(self.cdfg_as, inputs["graph"])

     # GNN
     # input layer -> n - 1 GNN layers (relu) -> final GNN layer (softmax)
diff --git a/nn/train.py b/nn/train.py
index 8f6dcb2..20114d2 100644
--- a/nn/train.py
+++ b/nn/train.py
@@ -243,6 +243,19 @@ def run(graph_dir, tf_data_dir, ckpt_dir=None, ckpt_name="best.ckpt",
   for k, ds in dataset.items():
     if ds:
       dataset[k] = ds.batch(batch_size)
+  model = get_d2v_model(**model_config)
+  for elem in dataset["test"].as_numpy_iterator():
+    elem = elem[0]
+    for k, v in elem.items():
+      elem[k] = tf.constant(v, dtype=v.dtype)
+    x = elem["test_parameters"]
+    with tf.GradientTape() as g:
+      g.watch(x)
+      y = model(elem)
+    dy_dx = g.gradient(y, x)
+    print(x)
+    print(dy_dx)
+    assert 0

   # Train model
   model = get_d2v_model(**model_config)
```

**Comment by insoochung** [comment]
@AaronGChan,

You'll also need to train and pass the model I guess. There is a way that you can achieve it without having to pass the whole model flags: https://www.tensorflow.org/tutorials/keras/save_and_load, but do start with whatever works for you. Also, let me know if you have trouble setting things up in grace, we could work it out through Teams.

And, could you give me an estimate time? Thanks.

**Comment by aarongchan** [comment]
I think by next Wednesday or Thursday I should be done with this task. I have a midterm next Tuesday, so I will be studying for it but will try to fit in time to work on this.
